core/src/connection_reuse.rs

```diff
@@ -40,193 +40,541 @@
 use upgrade::ConnectionUpgrade;
+use std::clone::Clone;
 
 /// Allows reusing the same muxed connection multiple times.
 ///
 /// Can be created from an `UpgradedNode` through the `From` trait.
 ///
-/// Implements the `Transport` trait.
+/// Implements the `Transport` trait
 #[derive(Clone)]
 pub struct ConnectionReuse<T, C>
 where
     T: Transport,
-    T::Output: AsyncRead + AsyncWrite,
+    C: ConnectionUpgrade<T::Output, T::MultiaddrFuture> + Clone,
+    C::Output: StreamMuxer + Clone,
+{
+    /// Struct shared between most of the `ConnectionReuse` infrastructure.
+    manager: Arc<ConnectionsManager<T, C>>,
+}
+
+/// Keeps an internal counter, for every new number issued
+/// will increase its internal state.
+#[derive(Default)]
+struct Counter {
+    inner: AtomicUsize,
+}
+
+impl Counter {
+    /// Returns the next counter, increasing the internal state
+    pub fn next(&self) -> usize {
+        self.inner.fetch_add(1, Ordering::Relaxed)
+    }
+}
+
+enum PeerState<M>
+where
+    M: StreamMuxer + Clone,
+{
+    /// Connection is active and can be used to open substreams.
+    Active {
+        /// The muxer to open new substreams.
+        muxer: M,
+        /// Id of the listener that created this connection, or `None` if it was opened by a
+        /// dialer.
+        listener_id: Option<usize>,
+        /// whether this connection was closed
+        closed: Arc<AtomicBool>
+    },
+
+    /// Connection is pending.
+    // TODO: stronger Future type
+    Pending {
+        /// Future that produces the muxer.
+        future: Box<Future<Item = (M, Multiaddr), Error = IoError>>,
+        /// All the tasks to notify when `future` resolves.
+        notify: Vec<(usize, task::Task)>,
+    },
+
+    /// An earlier connection attempt errored.
+    Errored(IoError),
+}
+
+/// Struct shared between most of the `ConnectionReuse` infrastructure.
+/// Knows about all connections and their current state, allows one to poll for
+/// incoming and outbound connections, while managing all state-transitions that
+/// might occur automatically.
+// #[derive(Clone)]
+struct ConnectionsManager<T, C>
+where
+    T: Transport,
     C: ConnectionUpgrade<T::Output, T::MultiaddrFuture>,
-    C::Output: StreamMuxer,
+    C::Output: StreamMuxer + Clone,
 {
-    // Underlying transport and connection upgrade for when we need to dial or listen.
-    inner: UpgradedNode<T, C>,
+    /// Underlying transport and connection upgrade, used when we need to dial or listen.
+    transport: UpgradedNode<T, C>,
+
+    /// All the connections that were opened, whether successful and/or active or not.
+    // TODO: this will grow forever
+    connections: Mutex<FnvHashMap<Multiaddr, PeerState<C::Output>>>,
 
-    // Struct shared between most of the `ConnectionReuse` infrastructure.
-    shared: Arc<Mutex<Shared<C::Output>>>,
+    /// Tasks to notify when one or more new elements were added to `connections`.
+    notify_on_new_connec: Mutex<Vec<(usize, task::Task)>>,
```

Comment:
We were wondering about the performance of this, so [I made a bunch of test](https://gist.github.com/gnunicorn/24d513debae25bfd3586e01e58f7dbd0). They showed that the overhead of hashing is only cut of once the `vec` is actually quite large (>200) - which is super unlikely to happen for the notifications (as barely ever more than a handfull of tasks would want to be notified, and we drain it every time we do). What makes the HashMap worse in this scenario, is that we "intend-to-insert" much more often than we actually insert, and every time we do, we do the expensive hashing in the hashmap while we'd exit early in looping through the vector. In practice the vector has one or two entries, making it faster to loop than to hash (even with `Fnv`). I was told these supposedly tiny performance differences matter, so I made the change.

Reply:
I do not think these benchmarks do the same thing.  For instance [`bench_custom_hasher_hashmap_same`](https://gist.github.com/gnunicorn/24d513debae25bfd3586e01e58f7dbd0#file-fnv_bench-rs-L77) does `COUNT` inserts, whereas [`bench_with_vec_same`](https://gist.github.com/gnunicorn/24d513debae25bfd3586e01e58f7dbd0#file-fnv_bench-rs-L165) returns early, i.e. after `MODULO` inserts. Changing `bench_with_vec_same` to do (roughly) what the `HashMap` based benchmark does


```rust
#[bench]
fn bench_with_vec_same(b: &mut Bencher) {
    b.iter(|| {
        let n = test::black_box(COUNT);
        let mut v: Vec<(Key, &'static str)> = vec![];
        'outer: for idx in 0..n {
            let id = idx % MODULO;
            for i in 0..v.len() {
                if v.get(i).unwrap().0 == id {
                    v[i] = (id, "value");
                    continue 'outer;
                }
            }
            v.push((id, "value"));
        };
    });
}
```

yields

```
running 2 tests
test bench_custom_hasher_hashmap ... bench:       8,800 ns/iter (+/- 376)
test bench_with_vec_same         ... bench:      32,970 ns/iter (+/- 1,188)
```

Reply:
@twittner oh, yes, what a stupid mistake - I had an iterator in there first and never replaced that `return`. I have to run these tests again.

why are you doing a `v[i] = (id, "value");` if you found it though?

Reply:
Sorry about that, that was a stupid bug. However, I inserted the `Modulo` only to emulate a scenario closer to our "many attempts little inserts" scenario. I only tried that because the worst-case-scenario (`never matches, inserts every time`) yielded good results for smaller sizes, which was unaffected by the bug. I fixed the code and running it now on [small sets still yields significant performance improvemnets](https://gist.github.com/gnunicorn/24d513debae25bfd3586e01e58f7dbd0#file-_results-txt-L2). On my system - take it with a grain of salt, flexible cpu-rate and all - the `HashMap` only shows benefit after 30+ items, which we'll probably not reach in practice:

```
# 500 Modulo 30
test bench_custom_hasher_hashmap_same ... bench:       5,626 ns/iter (+/- 608)
test bench_fnv_hashmap_same           ... bench:       8,043 ns/iter (+/- 3,692)
test bench_fnv_hashset_same           ... bench:       6,652 ns/iter (+/- 394)
test bench_fnv_hashset_uniques        ... bench:      12,490 ns/iter (+/- 2,910)
test bench_with_sorted_vec_same       ... bench:      11,130 ns/iter (+/- 457)
test bench_with_vec_same              ... bench:       6,454 ns/iter (+/- 154)

test result: ok. 0 passed; 0 failed; 0 ignored; 6 measured; 0 filtered out

```

Reply:
I can not reproduce those results. `bench_custom_hasher_hashmap_same` is always faster or on par with `bench_with_vec_same` on my machine, so I would prefer to go with Big-O and use a `HashMap`.

```
# 100 Modulo 20
running 2 tests
test bench_custom_hasher_hashmap_same ... bench:         909 ns/iter (+/- 62)
test bench_with_vec_same              ... bench:       1,026 ns/iter (+/- 59)

# 200 Modulo 20
running 2 tests
test bench_custom_hasher_hashmap_same ... bench:       1,782 ns/iter (+/- 74)
test bench_with_vec_same              ... bench:       1,903 ns/iter (+/- 93)

# 2000 Modulo 20
running 2 tests
test bench_custom_hasher_hashmap_same ... bench:      17,063 ns/iter (+/- 366)
test bench_with_vec_same              ... bench:      17,590 ns/iter (+/- 933)

# 1000 Modulo 100
running 2 tests
test bench_custom_hasher_hashmap_same ... bench:       9,126 ns/iter (+/- 448)
test bench_with_vec_same              ... bench:      38,695 ns/iter (+/- 1,522)

# 500 Modulo 30
running 2 tests
test bench_custom_hasher_hashmap_same ... bench:       4,605 ns/iter (+/- 237)
test bench_with_vec_same              ... bench:       7,246 ns/iter (+/- 471)
```

Reply:
@twittner and then use the custom hasher? Fine with me. I am not sure my results aren't influenced because of the automatic CPU-frequency-govener that I can't disable :(  . Especially for the first tests this probably skewes the result as it will only ramp up CPU freq during the test... That the custom hasher (which does nothing) is supposedly magnitutes slower than the Fnv supports that thesis - because Fnv doesn't do anything else than we do...

I'll do that.

Reply:
I think `SmallVec` should be faster because of cache locality.